src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -5957,20 +5957,62 @@ bool MergeTreeData::removeDetachedPart(DiskPtr disk, const String & path, const
     return false;
 }
 
-PartitionCommandsResultInfo MergeTreeData::unfreezePartitionsByMatcher(MatcherFn matcher, const String & backup_name, ContextPtr)
+// copy and some edits from StorageReplicatedMergeTree::removeSharedDetachedPart
+bool removeSharedDetachedPart(DiskPtr disk, const String & path, const String & part_name, const String & table_uuid,
+    const String &, const String & detached_replica_name, const String & detached_zookeeper_path, Poco::Logger * log, ContextPtr local_context)
```

Comment:
This code (`removeSharedDetachedPart`, `removeDetachedPart` and may be `unfreezePartitionsFromTableDirectory`) should be moved in some new separate class, I think. `FreezeMetaData` struct too.
Non-replicated `MergeTreeData` should know nothing about ZooKeeper.